enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,322 @@
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy turns on compression for the specified MIME types
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the specified
+// MIME types
+type HTTPCompressionPolicy struct {
+	// compressionMIMETypes is a list of MIME types that should have compression applied.
+	// At least one MIME type must be specified.
+	//
+	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
+	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
+	// formats benefit from compression, but formats that are already compressed (pdf, image,
+	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
+	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`
+}
+
+// CompressionMIMEType defines the format of a single MIME type
+// E.g. "text/css; charset=utf-8", "text/html", "text/*", "image/svg+xml",
+// "application/octet-stream", etc.
+//
+// +kubebuilder:validation:Pattern="[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"
+type CompressionMIMEType string
+```
+
+To set up the IngressControllerSpec to use HTTP compression for a set of specific MIME types, the configuration
+might resemble:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpCompression:
+    compressionMIMETypes:
+    - "text/html"
+    - "text/css; charset=utf-8"
+    - "application/json"
+```
+
+To expose the HTTP compression variables, further changes will be added to the ingressi router deployment:
+```go
+const (
+	WildcardRouteAdmissionPolicy = "ROUTER_ALLOW_WILDCARD_ROUTES"
+    ...
+	RouterEnableCompression = "ROUTER_ENABLE_COMPRESSION"
+	RouterCompressionMIMETypes = "ROUTER_COMPRESSION_MIME"
+)
+...
+    // desiredRouterDeployment returns the desired router deployment.
+    func desiredRouterDeployment(ci *operatorv1.IngressController, ingressControllerImage string, ingressConfig *configv1.Ingress, apiConfig *configv1.APIServer, networkConfig *configv1.Network, proxyNeeded bool, haveClientCAConfigmap bool, clientCAConfigmap *corev1.ConfigMap) (*appsv1.Deployment, error) {
+    ...
+      if len(ci.Spec.HTTPCompression.CompressionMIMETypes) != 0 {
+        env = append(env, corev1.EnvVar{Name: RouterEnableCompression, Value: "true"})
+        env = append(env, corev1.EnvVar{Name: RouterCompressionMIMETypes, Value: string.Join(spec.httpCompression.compressionMimeTypes, " ")})
```

Comment:
Do we need to quote or escape values with spaces (such as `text/css; charset=utf-8` in the example above)?

Reply:
The quotes need to be there already.  No validation done at this point.  I'll add quotes to the pattern.

Reply:
They need to be quoted, but I would leave that up to the user.  Wdyt?